packages/reference/src/common/customCardTypes.ts

```diff
@@ -0,0 +1,34 @@
+import { Asset, ContentType, Entry } from '../types';
+import * as React from 'react';
+import { CustomActionProps } from './ReferenceEditor';
+
+export type DefaultCardRenderer<T extends CustomCardProps> = (
+  props?: T
+) => React.ReactElement;
+
+export type CustomCardRenderer<T extends CustomCardProps> = (
+  props: T,
+  linkActionsProps: CustomActionProps,
+  renderDefaultCard: DefaultCardRenderer<T>
+) => React.ReactElement | false;
+
+export type CustomCardProps = {
+  index?: number;
+  localeCode: string;
+  defaultLocaleCode: string;
+  isDisabled: boolean;
+  size: 'default' | 'small';
+  cardDragHandle?: React.ReactElement;
+  onEdit?: () => void;
+  onRemove?: () => void;
+  onMoveTop?: () => void;
+  onMoveBottom?: () => void;
+  // TODO: specific types should be used instead of these props, but for that the ReferenceEditor props should accept a generic
+  // Entry editor
+  entry?: Entry;
```

Comment:
Would it be difficult to address above TODO and just make this `entity` instead of `entry?` and `asset?` ?

Reply:
I was trying to do that, but it all requires making the [`ReferenceEditorProps`](https://github.com/contentful/field-editors/blob/master/packages/reference/src/common/ReferenceEditor.tsx#L11-L35) generic. This will break many other things relying on this type. 

I also tried to make `ReferenceEditorProps.renderCustomCard<CustomEntryCardProps | CustomAssetCardProps>` and make this function type more specific in each of the editors types: `SingleEntryReferenceEditor`, `MultipleEntryReferenceEditor`, `MultipleMediaEditor` and `SingleMediaEditor`. But it still conflicts because of hierarchy and composing of the multiple components.

So my answer is that it's not easy unless I miss an obvious and simple approach.

Reply:
>  it all requires making the ReferenceEditorProps generic. 

They seem pretty generic already, with the exception of `renderCustomCard?.props: CustomEntryCardProps`. And that's because `CustomEntryCardProps` was only relevant to entry link (reference) editors so far.

So it all comes back to [this TODO](https://github.com/contentful/field-editors/blob/fcae1d5dc1423df1b4c4b346cdeee3ed9f2e6340/packages/reference/src/common/ReferenceEditor.tsx#L39-L41):
https://github.com/contentful/field-editors/blob/fcae1d5dc1423df1b4c4b346cdeee3ed9f2e6340/packages/reference/src/common/ReferenceEditor.tsx#L39-L43

But I guess that's nothing new for you, just trying to get an overview here.

> I also tried to make ReferenceEditorProps.renderCustomCard<CustomEntryCardProps | CustomAssetCardProps> and make this function type more specific in each of the editors types

Feels to me like a generic `CustomEntityCardProps` (or `CustomCardProps`) would be enough for now. But maybe I am missing something.

> This will break many other things relying on this type.
I think a major version bump and telling users to adjust to the new interfaces would be preferable over having less clean interfaces that might grow even more into this direction with further changes.

It's hard for me to keep an overview, but wouldn't we need to introduce a major version bump anyways considering your change of `CustomEntryCardProps` to `CustomCardProps`?